utils/sharedcache/common.go

```diff
@@ -0,0 +1,312 @@
+package sharedcache
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/collection"
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	"github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/hashing"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+	"github.com/ARM-software/golang-utils/utils/reflection"
+)
+
+const defaultCachedPackage = "cache.zip"
+const hashFileDescriptor = ".hash"
+
+// GenerateKey generates a key based on a list of key elements `elems`.
+func GenerateKey(elems ...string) string {
+	hash := "ef46db3751d8e999" // xxhash(""). No point in calculating it for "" everytime
+	for _, elem := range elems {
+		hash = hashing.CalculateHash(hash+elem, hashing.HashXXHash)
+	}
+	return hash
+}
+
+func generateHashFileName(srcFile string) string {
+	return fmt.Sprintf("%v%v", filepath.Base(srcFile), hashFileDescriptor)
+}
+func getHash(ctx context.Context, fs filesystem.FS, src string, forceHashUpdate bool) (hash string, err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	hashFile := filepath.Join(filepath.Dir(src), generateHashFileName(src))
+	if fs.Exists(hashFile) && !forceHashUpdate {
+		fileContents, readErr := fs.ReadFile(hashFile)
+		if readErr == nil {
+			hash = string(fileContents)
+			if len(hash) == 16 { // if valid hash
+				// xxhash 64 bit will have 16 characters
+				return
+			}
+		}
+	}
+
+	hash, err = fs.FileHash(hashing.HashXXHash, src)
+	if err != nil {
+		return
+	}
+	_ = fs.WriteFile(hashFile, []byte(hash), 0775)
+	return
+}
+
+func isPathForDirectory(fs filesystem.FS, dst string) bool {
+	if ok, _ := fs.IsDir(dst); ok {
+		return true
+	}
+	if filepath.Ext(dst) == "" {
+		return true
+	}
+	return strings.HasSuffix(dst, string(fs.PathSeparator())) || strings.HasSuffix(dst, "/")
+}
+
+// TransferFiles transfers a file from a location `src` to another `dest` and ensures the integrity (i.e. hash validation) of what was copied across.
+// `dest` can be a file or a directory. If not existent, it will be created on the fly. non-existent directory path should be terminated by a path separator i.e. / or \
+func TransferFiles(ctx context.Context, fs filesystem.FS, dst, src string) (destFile string, err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+	if !fs.Exists(src) {
+		err = fmt.Errorf("%w: source path does not exist [%v]", commonerrors.ErrNotFound, src)
+		return
+	}
+	if result, suberr := fs.IsFile(src); !result || suberr != nil {
+		err = fmt.Errorf("%w: source is not a file [%v, (%v)]", commonerrors.ErrInvalid, src, suberr)
```

Comment:
subErr could be an issue with permissions (it calls stat) or something, not just that the source is not a file so I think that the error could be improved